### DISCUSSION TITLE
SSL3 error when executing curl command with tls1.2 option

### DISCUSSION THREAD

**Discussion by kamiryo** [curl/curl#11100] [Q&A]
I would like to know why I get the curl error output "curl: (35) error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure".

The command executed is as follows
`curl --tlsv1.2 --cert xxx.pem --key xxxx.key https://example.com -x http://〇〇.〇〇.〇〇:8080 -X POST  -H "Content-Type: text/plain"  -vv`

` * error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure`
`* Closing connection 0`
`curl: (35) error:14094410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure`

I have obtained tcpdump and confirmed that the communication was carried out by tls1.2.
The direct cause of the error was that the server sent an allert message "handshake_failure(40)" to the client when we checked the tcpdump.The conclusion was an error in the client certificate pem file.

I think that the message that curl is outputting is not in alignment with Alert Messeage.

Best regards.

**Comment by jay** [comment]
> The direct cause of the error was that the server sent an allert message "handshake_failure(40)" to the client when we checked the tcpdump.The conclusion was an error in the client certificate pem file.
> 
> I think that the message that curl is outputting is not in alignment with Alert Messeage.

What version of curl are you using (`curl -V`) and how do you suggest we improve? We are passing on what OpenSSL says and it didn't say there was an error in the client certificate file.

https://github.com/curl/curl/blob/b16d1fa8ee567b52c09a0f89940b07d8491b881d/lib/vtls/openssl.c#L1287-L1303

Also, the alert "handshake failure" is the same as the error message so how are they not in alignment?

**Reply by kamiryo** [reply]
Thanks all for the lighting fast response!

> What version of curl are you using (curl -V) 

Version information is here.
```
curl 7.61.1 (x86_64-redhat-linux-gnu) libcurl/7.61.1 OpenSSL/1.1.1k zlib/1.2.11 brotli/1.0.6 libidn2/2.2.0 libpsl/0.20.2 (+libidn2/2.2.0) libssh/0.9.6/openssl/zlib nghttp2/1.33.0
Release-Date: 2018-09-05
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz brotli TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL
```

I don't understand the internal communication from openssl to curl, but what I find strange is that I am using TLS1.2 to make the connection, but as a curl command, I am getting the standard output on the console: sslv3 alert handshake failure. I am not sure what to do with this.

Also, the ssl Alert on tcpdump is 40, which I assume is RFC 5246 handshake_failure.
>Reception of a handshake_failure alert message indicates that the
sender was unable to negotiate an acceptable set of security
parameters given the options available. This is a fatal error.

Based on the content here, I thought it would be confusing to list it as sslv3 alert handshake failure,

**Reply by jay** [reply]
> Based on the content here, I thought it would be confusing to list it as sslv3 alert handshake failure

openssl/openssl#7953 fixed by openssl/openssl#17161

https://github.com/openssl/openssl/blob/OpenSSL_1_1_1k/ssl/ssl_err.c#L1104-L1105

https://github.com/openssl/openssl/blob/81b741f6/ssl/ssl_err.c#L400-L401
